### ISSUE TITLE
browse.path is utterly frustrating

### ISSUE THREAD

**Issue by Gruntfuggly** [microsoft/vscode-cpptools#1185]
- sometimes it can't find files in the workspace folder (ie, the same folder!)
- there seems to be no logic as to which paths it uses
- is it recursive or not?
- if it is, how many levels does it traverse?
- does it follow sym links?
- does it expand '~'?
- is the order of paths important?

Is there any documentation on how it should be used?

It seemed to working fine for me a week ago, but now it's stopped working...

**Comment by bobbrow** [comment]
Our documentation for c_cpp_properties.json is here: https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/c_cpp_properties.json.md

> - sometimes it can't find files in the workspace folder (ie, the same folder!)

please share your browse.path (if you can)

> - there seems to be no logic as to which paths it uses

if you [enable logging in the output window](https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/Enabling%20logging.md) (use the "Information" setting) and then reload the window we will print out what paths we are searching.

> - is it recursive or not?

it depends on whether the path ends with `*` or not.  The documentation above explains the behavior.

> - if it is, how many levels does it traverse?

looking at the code, I see we have a hard cap on 20 levels

> - does it follow sym links?

yes, it should. If it does not, that is a bug.

> - does it expand '~'?

yes. It also expands environment variables and the special `${workspaceRoot}` variable

> - is the order of paths important?

no

**Comment by Gruntfuggly** [comment]
Thanks for a comprehensive answer Bob. That gives me plenty of ammo to try and sort it out tomorrow.

Incidentally, is there a way I could have found that documentation? Working backwards, I can see I can get there from the github repo, but maybe there should be a link to it from the extension page?

**Comment by bobbrow** [comment]
Our release notes landing page has a link to it, but I see that the marketplace page links to a different location.  We should sort this out.  Thanks for bringing this to our attention.

**Comment by Gruntfuggly** [comment]
I've tried turning on the logging as suggested, but C_cpp.loggingLevel seems to be unrecognised, and there was no extra debug in the output window.

My browse path is currently:
```
            "browse": {
                "path": [
                    "/usr/include/c++/5",
                    "/usr/include/x86_64-linux-gnu/c++/5",
                    "/usr/local/include",
                    "/usr/include",
                    "/usr/include/x86_64-linux-gnu",
                    "${workspaceRoot}",
                    "~/workspace/tmb"
                ],
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            }
```
workspaceRoot is `/home/nscott/workspace/tmb/daemons/inpddud`
With it set like this, it is not finding included files such as 
```
/home/nscott/workspace/tmb/dataModel/baseconfig.h
/home/nscott/workspace/tmb/tmm/libs/tmmcomm/tmmcomm.h
```

Using Insiders, it can't find
```
/home/nscott/workspace/tmb/daemons/inpddud/inpddud-treeitemtoplevel.h
```

But oddly, it can find my Qt headers, which are in 
```
/usr/include/qt4
```

I have tried resetting the database and manually deleting it, before restarting.

There is one potential issue in that there is a symlinked folder in `/home/nscott/workspace/tmb` which has lots of nested sub directories. Is it possible that it is giving up while indexing this folder?

**Comment by bobbrow** [comment]
Did you check the C/C++ channel in the Output window?  It may not have been auto-selected by default.

![image](https://user-images.githubusercontent.com/12818240/32239089-d008b274-be26-11e7-8a6c-5f68d2fd43e5.png)

**Comment by Gruntfuggly** [comment]
Didn't have the latest version of the extension installed (sorry).

The output shows:
```
IntelliSense Engine = Tag Parser.
The extension will use the Tag Parser for IntelliSense when #includes don't resolve.
Autocomplete is enabled.
Error squiggles are enabled.
File exclude: **/.git
File exclude: **/.svn
File exclude: **/.hg
File exclude: **/CVS
File exclude: **/.DS_Store
File exclude: **/gen
File exclude: **/Entries
File exclude: **/Makefile*
File exclude: **/moc_*
File exclude: **/bin
File exclude: **/.obj
File exclude: **/.moc
File exclude: **/libs
File exclude: **/GPATH
File exclude: **/GRTAGS
File exclude: **/GTAGS
File exclude: **/ivencs-sdk*
File exclude: **/tmm/asl0632
File exclude: /home/nscott/workspace/tmb/daemons/tmm/include/dataModel/*
File exclude: **/.vscode
Search exclude: **/node_modules
Search exclude: **/bower_components
Search exclude: **/*.mak
Search exclude: **/*.min.js
Search exclude: **/catch.hpp
Search exclude: **/.vscode
Code browsing service initialized
  Folder: /usr/local/include/ will be indexed
  Folder: /usr/include/ will be indexed
  Folder: /home/nscott/workspace/tmb/ will be indexed
Discovering files...
  Processing folder (recursive): /usr/local/include/
  Processing folder (recursive): /usr/include/
  Processing folder (recursive): /home/nscott/workspace/tmb/
  Indexing: 7485 file(s) processed
  Indexing: 0 file(s) removed from database
Done discovering files.
Parsing open files...
Checking for edits and parsing remaining files...
  Indexing: 0 file(s) processed
Done parsing remaining files.
Tag parsing failed: /usr/include/x86_64-linux-gnu/bits/mathinline.h
Done parsing open files.
```

I tried excluding the file it says it failed on, but it doesn't help.

**Comment by bobbrow** [comment]
I see one potential problem.  This is a regression in the extension that was recently introduced.  Your browse.path is including a path that is a parent of `${workspaceRoot}`.  We recently updated the extension to avoid scanning the same folder more than once and we merge paths when possible.  It turns out that this breaks another section of code that turns off indexing for directories not under `${workspaceRoot}` (the check is done at the top level folder, not at each folder during the recursion).  This results in us parsing nothing in your workspace, thus the `Indexing: 0 file(s) processed` in your log.

As a workaround until we fix this, you would need to change your `~/workspace/tnb` browse.path entry to not recursively include your inpddud folder. e.g. use `~/workspace/tmb/dataModel`, `~/workspace/tmb/tmm`, and `~/workspace/tmb/daemons/*` instead of `~/workspace/tmb`

But this issue may not be the cause of all your woes since we do track all of the files, and whenever you open a file we will parse it and its #include dependencies.  So there may be something else here too.  What exactly do you mean when you say that it can't find `/home/nscott/workspace/tmb/daemons/inpddud/inpddud-treeitemtoplevel.h`?  Which operation is failing?  Go to definition?  Auto-completion?  Something else?

**Comment by Gruntfuggly** [comment]
I've managed to get it working by removing the recursive folder `~/workspace/tmb` and adding the subfolders individually.

When it couldn't find `/home/nscott/workspace/tmb/daemons/inpddud/inpddud-treeitemtoplevel.h` it would display squiggles under the #include statement, and I could visit definitions contained in it.

Thanks for your explanation and work around. Bizarrely it only affects some of my projects - some seem to fine with the recursive folder.

**Comment by bobbrow** [comment]
We have a fix for the browse.path bug and it will be out with our next release (expected to be this week).

I'm still unclear as to what the problem with `inpddud-treeitemtoplevel.h` is.  Is the squiggle in `inpddud-treeitemtoplevel.h` or in some other file that includes `inpddud-treeitemtoplevel.h`?  Were you able to navigate from some other file into `inpddud-treeitemtoplevel.h`, or did the navigation originate from `inpddud-treeitemtoplevel.h` and just go to another symbol in the same file?

**Comment by Gruntfuggly** [comment]
It is included by other files, including `inpddud-treeitemtoplevel.cpp`, which is in the same folder. The squiggles are under the include statement, so I couldn't navigate to it from the other files.

**Comment by sean-mcmanus** [comment]
Should have been fixed with 0.14.1.
